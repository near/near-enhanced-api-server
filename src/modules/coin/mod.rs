use actix_web::{HttpResponse, error};
use actix_web_validator::{Error, PathConfig};
use paperclip::actix::web;

mod data_provider;
mod resources;
mod schemas;

#[derive(serde::Serialize)]
pub struct ValidationErrorJsonPayload {
    pub message: String,
    pub fields: Vec<String>,
}

/// Custom error handler
impl From<&validator::ValidationErrors> for ValidationErrorJsonPayload {
    fn from(error: &validator::ValidationErrors) -> Self {
        ValidationErrorJsonPayload {
            message: "Validation error".to_owned(),
            fields: error.field_errors().iter().map(|(field, _)| field.to_string()).collect(),
        }
    }
}

pub(crate) fn register_services(app: &mut web::ServiceConfig) {
    app
        .app_data(PathConfig::default().error_handler(|err, _| {
            let json_error = match &err {
                Error::Validate(error) => ValidationErrorJsonPayload::from(error),
                _ => ValidationErrorJsonPayload { message: err.to_string(), fields: Vec::new() },
            };
            error::InternalError::from_response(err, HttpResponse::Conflict().json(json_error)).into()
        }))
        .service(
        web::resource("/accounts/{account_id}/coins/NEAR")
            .route(web::get().to(resources::get_near_balance)),
    )
    .service(
        web::resource("/accounts/{account_id}/coins")
            .route(web::get().to(resources::get_coin_balances)),
    )
    .service(
        web::resource("/accounts/{account_id}/coins/{contract_account_id}")
            .route(web::get().to(resources::get_coin_balances_by_contract)),
    )
    .service(
        web::resource("/accounts/{account_id}/coins/NEAR/history")
            .route(web::get().to(resources::get_near_history)),
    )
    .service(
        web::resource("/accounts/{account_id}/coins/{contract_account_id}/history")
            .route(web::get().to(resources::get_coin_history)),
    )
    .service(
        web::resource("/nep141/metadata/{contract_account_id}")
            .route(web::get().to(resources::get_ft_contract_metadata)),
    );
}
