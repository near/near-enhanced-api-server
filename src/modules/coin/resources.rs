use paperclip::actix::{
    api_v2_operation,
    web::{self, Json},
};

use super::{data_provider, schemas};
use crate::{db_helpers, errors, modules, types};

#[api_v2_operation(tags(Coins))]
/// Get user's NEAR balance
///
/// This endpoint returns the NEAR balance of the given `account_id`
/// at the given `block_timestamp_nanos`/`block_height`.
pub async fn get_near_balance(
    pool: web::Data<sqlx::Pool<sqlx::Postgres>>,
    rpc_client: web::Data<near_jsonrpc_client::JsonRpcClient>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::BalanceRequest>,
    block_params: web::Query<types::query_params::BlockParams>,
) -> crate::Result<Json<schemas::NearBalanceResponse>> {
    let block = db_helpers::checked_get_block(&pool, &block_params).await?;
    modules::check_account_exists(&rpc_client, &request.account_id.0, block.height).await?;

    Ok(Json(
        data_provider::get_near_balance(&pool, &block, &request.account_id.0).await?,
    ))
}

#[api_v2_operation(tags(Coins))]
/// Get user's coin balances
///
/// This endpoint returns all the countable coin balances (including NEAR, FTs)
/// of the given `account_id`, at the given `block_timestamp_nanos`/`block_height`.
/// todo it's not great, we can't keep sorting + pagination when we add MT here
/// NEAR balance always goes first, then the balances are sorted alphabetically by `contract_account_id`.
///
/// You can also provide `limit` (max 100) and `after_contract_account_id`.
/// `after_contract_account_id` is the way to paginate through the values.
///
/// Some pagination notes: leave `after_contract_account_id` empty at your first request;
/// then, take `contract_account_id` of the last provided item in your previous response.
///
/// **Limitations**
/// * We are in the process of supporting Multi Token balances.
pub async fn get_coin_balances(
    pool: web::Data<sqlx::Pool<sqlx::Postgres>>,
    rpc_client: web::Data<near_jsonrpc_client::JsonRpcClient>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::BalanceRequest>,
    block_params: web::Query<types::query_params::BlockParams>,
    limit_params: web::Query<types::query_params::LimitParams>,
) -> crate::Result<Json<schemas::CoinBalancesResponse>> {
    let mut limit = types::query_params::checked_get_limit(limit_params.limit)?;
    let block = db_helpers::checked_get_block(&pool, &block_params).await?;
    modules::check_account_exists(&rpc_client, &request.account_id.0, block.height).await?;

    let mut balances: Vec<schemas::Coin> = vec![];
    balances.push(
        data_provider::get_near_balance(&pool, &block, &request.account_id.0)
            .await?
            .into(),
    );
    limit -= 1;

    if limit > 0 {
        let ft_balances = &mut data_provider::get_coin_balances(
            &pool,
            &rpc_client,
            &block,
            &request.account_id.0,
            limit,
        )
        .await?;
        balances.append(ft_balances);
    }

    Ok(Json(schemas::CoinBalancesResponse {
        balances,
        block_timestamp_nanos: types::U64::from(block.timestamp),
        block_height: types::U64::from(block.height),
    }))
}

#[api_v2_operation(tags(Coins))]
/// Get user's coin balances by contract
///
/// This endpoint returns all the countable coin balances of the given `account_id`,
/// for the given `contract_account_id` and `block_timestamp_nanos`/`block_height`.
/// For FT contracts, the response has only 1 item in the list.
/// For MT contracts, there could be several balances (MT support is still under development).
/// todo change interface for ft/mt
///
/// **Limitations**
/// * We are in the process of supporting Multi Token balances.
pub async fn get_coin_balances_by_contract(
    pool: web::Data<sqlx::Pool<sqlx::Postgres>>,
    rpc_client: web::Data<near_jsonrpc_client::JsonRpcClient>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::BalanceByContractRequest>,
    block_params: web::Query<types::query_params::BlockParams>,
) -> crate::Result<Json<schemas::CoinBalancesResponse>> {
    if request.contract_account_id.to_string() == "near" {
        return Err(errors::ErrorKind::InvalidInput(
            "For native balance, please use NEAR (uppercase)".to_string(),
        )
        .into());
    }
    let block = db_helpers::checked_get_block(&pool, &block_params).await?;
    modules::check_account_exists(&rpc_client, &request.account_id.0, block.height).await?;

    let balances = data_provider::get_coin_balances_by_contract(
        &rpc_client,
        &block,
        &request.contract_account_id.0,
        &request.account_id.0,
    )
    .await?;

    Ok(Json(schemas::CoinBalancesResponse {
        balances,
        block_timestamp_nanos: types::U64::from(block.timestamp),
        block_height: types::U64::from(block.height),
    }))
}

#[api_v2_operation(tags(Coins))]
/// Get user's NEAR history
///
/// This endpoint returns the history of native NEAR operations,
/// for the given `account_id`, going from the latest to the earliest events.
/// For the next page, use `after_event_index` and pass there the index from the last event from the previous page.
///
/// **Limitations**
/// We are in the process of collecting the data. We support only last 3 months history for now.
pub async fn get_near_history(
    pool: web::Data<sqlx::Pool<sqlx::Postgres>>,
    pool_balances: web::Data<db_helpers::DBWrapper>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::BalanceRequest>,
    pagination_params: web::Query<types::query_params::PaginationParams>,
) -> crate::Result<Json<schemas::HistoryResponse>> {
    let pagination = modules::checked_get_pagination_params(&pagination_params).await?;
    let block = db_helpers::checked_get_block_from_pagination(&pool, &pagination).await?;
    // we don't need to check whether account exists. If not, we can just return the empty history

    Ok(Json(schemas::HistoryResponse {
        history: data_provider::get_near_history(
            &pool_balances.pool,
            &request.account_id.0,
            &block,
            &pagination,
        )
        .await?,
        block_timestamp_nanos: types::U64::from(block.timestamp),
        block_height: types::U64::from(block.height),
    }))
}

#[api_v2_operation(tags(Coins))]
/// Get user's coin history by contract
///
/// This endpoint returns the history of coin operations (FT, other standards)
/// for the given `account_id`, `contract_id`, going from the latest to the earliest events.
/// For the next page, use `after_event_index` and pass there the index from the last event from the previous page.
///
/// **Limitations**
/// * If the history does not match with the balances from RPC, we do not support the history (see Trusted/Inconsistent Contracts section below).
/// * We are in the process of supporting Multi Token history.
///
/// **Trusted/Inconsistent Contracts**
/// The contracts produce [the balance changing events](https://nomicon.io/Standards/Tokens/FungibleToken/Event).
/// The contract is marked as inconsistent if its events do not match with the balance taken from RPC.
/// The contract is marked as trusted otherwise.
/// More info could be found [here](https://github.com/near/near-indexer-events#my-contract-produces-eventstheres-a-custom-legacy-logic-for-my-contract-but-the-enhanced-api-still-ignores-me-why).
pub async fn get_coin_history(
    pool: web::Data<sqlx::Pool<sqlx::Postgres>>,
    pool_balances: web::Data<db_helpers::DBWrapper>,
    rpc_client: web::Data<near_jsonrpc_client::JsonRpcClient>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::HistoryRequest>,
    pagination_params: web::Query<types::query_params::PaginationParams>,
) -> crate::Result<Json<schemas::HistoryResponse>> {
    if request.contract_account_id.to_string() == "near" {
        return Err(errors::ErrorKind::InvalidInput(
            "For native history, please use NEAR (uppercase)".to_string(),
        )
        .into());
    }

    let pagination = modules::checked_get_pagination_params(&pagination_params).await?;
    let block = db_helpers::checked_get_block_from_pagination(&pool, &pagination).await?;
    // we don't need to check whether account exists. If not, we can just return the empty history

    Ok(Json(schemas::HistoryResponse {
        history: data_provider::get_coin_history(
            &pool,
            &pool_balances.pool,
            &rpc_client,
            &request.contract_account_id.0,
            &request.account_id.0,
            &block,
            &pagination,
        )
        .await?,
        block_timestamp_nanos: types::U64::from(block.timestamp),
        block_height: types::U64::from(block.height),
    }))
}

#[api_v2_operation(tags(Coins))]
/// Get FT contract metadata
///
/// This endpoint returns the metadata for a given FT contract and `timestamp`/`block_height`.
///
/// **Limitations**
/// * For now, we support only FT contracts that implement the Events NEP standard.
///   We are working on a solution to support other FT contracts, including `wrap.near` and bridged tokens.
pub async fn get_ft_contract_metadata(
    pool: web::Data<sqlx::Pool<sqlx::Postgres>>,
    rpc_client: web::Data<near_jsonrpc_client::JsonRpcClient>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::ContractMetadataRequest>,
    block_params: web::Query<types::query_params::BlockParams>,
) -> crate::Result<Json<schemas::FtContractMetadataResponse>> {
    let block = db_helpers::checked_get_block(&pool, &block_params).await?;

    Ok(Json(schemas::FtContractMetadataResponse {
        metadata: data_provider::get_ft_contract_metadata(
            &rpc_client,
            request.contract_account_id.0.clone(),
            block.height,
        )
        .await?,
        block_timestamp_nanos: types::U64::from(block.timestamp),
        block_height: types::U64::from(block.height),
    }))
}
