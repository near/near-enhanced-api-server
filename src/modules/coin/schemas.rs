use paperclip::actix::Apiv2Schema;
use validator::{Validate, ValidationError};

use crate::types;

// *** Requests ***

// move to coins
#[derive(Validate, Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct BalanceRequest {
    #[validate(custom = "near_primitives::types::AccountId::validate")]
    pub account_id: types::AccountId,
}

#[derive(Validate, Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct BalanceByContractRequest {
    #[validate(custom = "near_primitives::types::AccountId::validate")]
    pub account_id: types::AccountId,
    #[validate(custom = "near_primitives::types::AccountId::validate")]
    pub contract_account_id: types::AccountId,
}

#[derive(Validate, Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct HistoryRequest {
    #[validate(custom = "near_primitives::types::AccountId::validate")]
    pub account_id: types::AccountId,
    pub contract_account_id: types::AccountId,
}

// duplicate in each folder
#[derive(Validate, Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct ContractMetadataRequest {
    #[validate(custom = "near_primitives::types::AccountId::validate")]
    pub contract_account_id: types::AccountId,
}

// *** Responses ***

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct NearBalanceResponse {
    /// Sum of staked and nonstaked balances
    pub balance: types::U128,
    pub metadata: CoinMetadata,
    pub block_timestamp_nanos: types::U64,
    pub block_height: types::U64,
}

/// This response gives the information about all the available balances for the user.
/// The answer gives the list of NEAR, FT balances, could be used for Multi Tokens.
/// For MTs and other standards, balances could have multiple entries for one contract.
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct CoinBalancesResponse {
    pub balances: Vec<Coin>,
    pub block_timestamp_nanos: types::U64,
    pub block_height: types::U64,
}

/// This response provides the coin history (NEAR or by contract).
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct HistoryResponse {
    pub history: Vec<HistoryItem>,
    pub block_timestamp_nanos: types::U64,
    pub block_height: types::U64,
}

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct FtContractMetadataResponse {
    pub metadata: FtContractMetadata,
    pub block_timestamp_nanos: types::U64,
    pub block_height: types::U64,
}

// ---

/// This type describes general coin information.
/// It is used for NEAR, FT balances, could be used for Multi Tokens.
///
/// For MTs and other standards, we could have multiple coins for one contract.
/// For NEAR and FTs, coin_metadata contains general metadata (the only available option, though).
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct Coin {
    /// "nearprotocol" for NEAR, "nep141" for FT
    pub standard: String,
    pub balance: types::U128,
    /// null for NEAR, not null otherwise
    pub contract_account_id: Option<types::AccountId>,
    pub metadata: CoinMetadata,
    // TODO PHASE 1 (idea) I think it would be great to add here the info about last update moment. Timestamp, later also index
    // I'm already doing it at NftCount
}

/// This type describes the history of coin movements for the given user.
/// Coins could be NEAR, FT, it could be also later used for Multi Tokens.
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct HistoryItem {
    // TODO PHASE 2 add index here
    // pub index: types::U128,
    // TODO PHASE 1 (idea) do we want to add here tx_hash/receipt_id? We may want to add it at many places
    pub involved_account_id: Option<types::AccountId>,
    pub delta_balance: types::I128,
    pub balance: types::U128,
    pub cause: String,
    pub status: String,
    pub coin_metadata: CoinMetadata,
    pub block_timestamp_nanos: types::U64,
    // TODO PHASE 2 add this when we have all the data in the same DB. Now we can't join with blocks
    // pub block_height: types::U64,
}

/// This type describes general Metadata info, collecting the most important fields from different standards in the one format.
/// `decimals` may contain `0` if it's not applicable (e.g. if it's general MT metadata)
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct CoinMetadata {
    pub name: String,
    pub symbol: String,
    pub icon: Option<String>,
    pub decimals: u8,
}

/// The type for FT Contract Metadata. Inspired by
/// https://nomicon.io/Standards/Tokens/FungibleToken/Metadata
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct FtContractMetadata {
    pub spec: String,
    pub name: String,
    pub symbol: String,
    pub icon: Option<String>,
    pub reference: Option<String>,
    pub reference_hash: Option<String>,
    pub decimals: u8,
}

pub fn validate(account_id: &str) -> Result<(), ValidationError> {
    Err(ValidationError::new("something"))
}
