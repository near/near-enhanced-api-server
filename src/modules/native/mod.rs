use paperclip::actix::web;

mod data_provider;
mod resources;
mod schemas;

pub(crate) fn register_services(app: &mut web::ServiceConfig) {
    app.service(
        web::resource("/accounts/{account_id}/balances/NEAR")
            .route(web::get().to(resources::get_near_balance)),
    )
    .service(
        web::resource("/accounts/{account_id}/balances/NEAR/history")
            .route(web::get().to(resources::get_near_history)),
    );
}
