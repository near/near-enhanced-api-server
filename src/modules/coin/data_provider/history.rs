use std::str::FromStr;

use crate::modules::coin;
use crate::{db_helpers, errors, types};

// TODO PHASE 2 pagination by artificial index added to balance_changes
pub(crate) async fn get_near_history(
    balances_pool: &sqlx::Pool<sqlx::Postgres>,
    account_id: &near_primitives::types::AccountId,
    pagination: &types::query_params::HistoryPagination,
) -> crate::Result<Vec<coin::schemas::HistoryItem>> {
    let query = r"
        SELECT
            involved_account_id,
            delta_nonstaked_amount + delta_staked_amount delta_balance,
            absolute_nonstaked_amount + absolute_staked_amount balance,
            cause,
            status,
            block_timestamp block_timestamp_nanos
        FROM balance_changes
        WHERE affected_account_id = $1 AND block_timestamp < $2::numeric(20, 0)
        ORDER BY block_timestamp DESC
        LIMIT $3::numeric(20, 0)
    ";

    let history_info = db_helpers::select_retry_or_panic::<super::models::NearHistoryInfo>(
        balances_pool,
        query,
        &[
            account_id.to_string(),
            pagination.block_timestamp.to_string(),
            pagination.limit.to_string(),
        ],
    )
    .await?;

    let mut result: Vec<coin::schemas::HistoryItem> = vec![];
    for history in history_info {
        result.push(history.try_into()?);
    }
    Ok(result)
}

// TODO PHASE 2 pagination by artificial index added to assets__fungible_token_events
// TODO PHASE 2 change RPC call to DB call by adding absolute amount values to assets__fungible_token_events
// TODO PHASE 2 make the decision about separate FT/MT tables or one table. Pagination implementation depends on this
pub(crate) async fn get_coin_history(
    pool: &sqlx::Pool<sqlx::Postgres>,
    rpc_client: &near_jsonrpc_client::JsonRpcClient,
    contract_id: &near_primitives::types::AccountId,
    account_id: &near_primitives::types::AccountId,
    pagination: &types::query_params::HistoryPagination,
) -> crate::Result<Vec<coin::schemas::HistoryItem>> {
    // this is temp solution before we make changes to the DB
    let mut last_balance = super::balance::get_ft_balance_by_contract(
        rpc_client,
        contract_id.clone(),
        account_id.clone(),
        pagination.block_height,
    )
    .await?;
    let metadata = coin::schemas::CoinMetadata::from(
        super::metadata::get_ft_contract_metadata(
            rpc_client,
            contract_id.clone(),
            pagination.block_height,
        )
        .await?,
    );

    let account_id = account_id.to_string();
    let query = r"
        SELECT
            -- blocks.block_height,
            blocks.block_timestamp,
            assets__fungible_token_events.amount::numeric(45, 0),
            assets__fungible_token_events.event_kind::text cause,
            CASE WHEN execution_outcomes.status IN ('SUCCESS_VALUE', 'SUCCESS_RECEIPT_ID') THEN 'SUCCESS'
                ELSE 'FAILURE'
            END status,
            assets__fungible_token_events.token_old_owner_account_id old_owner_id,
            assets__fungible_token_events.token_new_owner_account_id new_owner_id
        FROM assets__fungible_token_events
            JOIN blocks ON assets__fungible_token_events.emitted_at_block_timestamp = blocks.block_timestamp
            JOIN execution_outcomes ON assets__fungible_token_events.emitted_for_receipt_id = execution_outcomes.receipt_id
        WHERE emitted_by_contract_account_id = $1
            AND (token_old_owner_account_id = $2 OR token_new_owner_account_id = $2)
            AND emitted_at_block_timestamp <= $3::numeric(20, 0)
        ORDER BY emitted_at_block_timestamp desc
        LIMIT $4::numeric(20, 0)
    ";
    let history_info = db_helpers::select_retry_or_panic::<super::models::CoinHistoryInfo>(
        pool,
        query,
        &[
            contract_id.to_string(),
            account_id.clone(),
            pagination.block_timestamp.to_string(),
            pagination.limit.to_string(),
        ],
    )
    .await?;

    let mut result: Vec<coin::schemas::HistoryItem> = vec![];
    for db_info in history_info {
        let mut delta: i128 = types::numeric::to_i128(&db_info.amount)?;
        let balance = last_balance;
        // TODO PHASE 2 maybe we want to change assets__fungible_token_events also to affected/involved?
        let involved_account_id = if account_id == db_info.old_owner_id {
            delta = -delta;
            types::account_id::extract_account_id(&db_info.new_owner_id)?
        } else if account_id == db_info.new_owner_id {
            types::account_id::extract_account_id(&db_info.old_owner_id)?
        } else {
            return Err(
                errors::ErrorKind::InternalError(
                    format!("The account {} should be sender or receiver ({}, {}). If you see this, please create the issue",
                            account_id, db_info.old_owner_id, db_info.new_owner_id)).into(),
            );
        };

        if db_info.status == "SUCCESS" {
            // TODO PHASE 2 this strange error will go away after we add absolute amounts to the DB
            if (last_balance as i128) - delta < 0 {
                return Err(errors::ErrorKind::InternalError(format!(
                    "Balance could not be negative: account {}, contract {}",
                    account_id, contract_id
                ))
                .into());
            }
            last_balance = ((last_balance as i128) - delta) as u128;
        }

        result.push(coin::schemas::HistoryItem {
            cause: db_info.cause.clone(),
            involved_account_id: involved_account_id.map(|id| id.into()),
            delta_balance: delta.into(),
            balance: balance.into(),
            coin_metadata: metadata.clone(),
            block_timestamp_nanos: types::numeric::to_u64(&db_info.block_timestamp)?.into(),
            // block_height: types::numeric::to_u64(&db_info.block_height)?.into(),
            status: db_info.status,
        });
    }
    Ok(result)
}

impl TryFrom<super::models::NearHistoryInfo> for coin::schemas::HistoryItem {
    type Error = errors::Error;

    fn try_from(info: super::models::NearHistoryInfo) -> crate::Result<Self> {
        let involved_account_id: Option<types::AccountId> =
            if let Some(account_id) = info.involved_account_id {
                Some(near_primitives::types::AccountId::from_str(&account_id)?.into())
            } else {
                None
            };
        Ok(Self {
            involved_account_id,
            delta_balance: types::numeric::to_i128(&info.delta_balance)?.into(),
            balance: types::numeric::to_u128(&info.balance)?.into(),
            cause: info.cause,
            status: info.status,
            coin_metadata: super::get_near_metadata(),
            block_timestamp_nanos: types::numeric::to_u64(&info.block_timestamp_nanos)?.into(),
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::modules::tests::*;

    #[tokio::test]
    async fn test_near_history() {
        let block = get_block();
        // Using the other pool because we have this table at the other DB
        dotenv::dotenv().ok();
        let url_balances =
            &std::env::var("DATABASE_URL_BALANCES").expect("failed to get database url");
        let pool = sqlx::PgPool::connect(url_balances)
            .await
            .expect("failed to connect to the balances database");
        let account = near_primitives::types::AccountId::from_str("vasya.near").unwrap();
        let pagination = types::query_params::HistoryPagination {
            block_height: block.height,
            block_timestamp: block.timestamp,
            limit: 10,
        };

        let balance = get_near_history(&pool, &account, &pagination).await;
        insta::assert_debug_snapshot!(balance);
    }

    #[tokio::test]
    async fn test_near_history_with_failed_receipts() {
        let block = db_helpers::Block {
            timestamp: 1618591017607373869,
            height: 34943083
        };
        // Using the other pool because we have this table at the other DB
        dotenv::dotenv().ok();
        let url_balances =
            &std::env::var("DATABASE_URL_BALANCES").expect("failed to get database url");
        let pool = sqlx::PgPool::connect(url_balances)
            .await
            .expect("failed to connect to the balances database");
        let account = near_primitives::types::AccountId::from_str("zubkowi.near").unwrap();
        let pagination = types::query_params::HistoryPagination {
            block_height: block.height,
            block_timestamp: block.timestamp,
            limit: 10,
        };

        let balance = get_near_history(&pool, &account, &pagination).await;
        insta::assert_debug_snapshot!(balance);
    }

    #[tokio::test]
    async fn test_coin_history() {
        let pool = init_db().await;
        let rpc_client = init_rpc();
        let block = get_block();
        let contract = near_primitives::types::AccountId::from_str("usn").unwrap();
        let account = near_primitives::types::AccountId::from_str("pushxo.near").unwrap();
        let pagination = types::query_params::HistoryPagination {
            block_height: block.height,
            block_timestamp: block.timestamp,
            limit: 10,
        };

        let balance = get_coin_history(&pool, &rpc_client, &contract, &account, &pagination).await;
        insta::assert_debug_snapshot!(balance);
    }

    #[tokio::test]
    async fn test_coin_history_with_failed_receipts() {
        let pool = init_db().await;
        let rpc_client = init_rpc();
        let block = db_helpers::Block {
            timestamp: 1651062637353692535,
            height: 64408633
        };
        let contract = near_primitives::types::AccountId::from_str("sweat_token_testing.near").unwrap();
        let account = near_primitives::types::AccountId::from_str("intmainreturn0.near").unwrap();
        let pagination = types::query_params::HistoryPagination {
            block_height: block.height,
            block_timestamp: block.timestamp,
            limit: 10,
        };

        let balance = get_coin_history(&pool, &rpc_client, &contract, &account, &pagination).await;
        insta::assert_debug_snapshot!(balance);
    }
}
