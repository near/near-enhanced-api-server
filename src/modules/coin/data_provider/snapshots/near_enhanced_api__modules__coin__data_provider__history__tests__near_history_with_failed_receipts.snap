---
source: src/modules/coin/data_provider/history.rs
expression: balance
---
Ok(
    [
        HistoryItem {
            involved_account_id: Some(
                AccountId(
                    "zubkowi.near",
                ),
            ),
            delta_balance: I128(
                -4114870673270849579420,
            ),
            balance: U128(
                31545068376206505050420580,
            ),
            cause: "TRANSACTION",
            status: "SUCCESS",
            coin_metadata: CoinMetadata {
                name: "NEAR blockchain native token",
                symbol: "NEAR",
                icon: Some(
                    "https://raw.githubusercontent.com/near/near-wallet/7ef3c824404282b76b36da2dff4f3e593e7f928d/packages/frontend/src/images/near.svg",
                ),
                decimals: 24,
            },
            block_timestamp_nanos: U64(
                1618591016599569765,
            ),
        },
        HistoryItem {
            involved_account_id: Some(
                AccountId(
                    "zubkowi.near",
                ),
            ),
            delta_balance: I128(
                79891276688900000000,
            ),
            balance: U128(
                31545148267483193950420580,
            ),
            cause: "CONTRACT_REWARD",
            status: "FAILURE",
            coin_metadata: CoinMetadata {
                name: "NEAR blockchain native token",
                symbol: "NEAR",
                icon: Some(
                    "https://raw.githubusercontent.com/near/near-wallet/7ef3c824404282b76b36da2dff4f3e593e7f928d/packages/frontend/src/images/near.svg",
                ),
                decimals: 24,
            },
            block_timestamp_nanos: U64(
                1618591016599569765,
            ),
        },
        HistoryItem {
            involved_account_id: None,
            delta_balance: I128(
                3362968354850349579420,
            ),
            balance: U128(
                31549183246879775900000000,
            ),
            cause: "RECEIPT",
            status: "SUCCESS",
            coin_metadata: CoinMetadata {
                name: "NEAR blockchain native token",
                symbol: "NEAR",
                icon: Some(
                    "https://raw.githubusercontent.com/near/near-wallet/7ef3c824404282b76b36da2dff4f3e593e7f928d/packages/frontend/src/images/near.svg",
                ),
                decimals: 24,
            },
            block_timestamp_nanos: U64(
                1618591001077258171,
            ),
        },
        HistoryItem {
            involved_account_id: Some(
                AccountId(
                    "zubkowi.near",
                ),
            ),
            delta_balance: I128(
                79891276688900000000,
            ),
            balance: U128(
                31545820278524925550420580,
            ),
            cause: "CONTRACT_REWARD",
            status: "FAILURE",
            coin_metadata: CoinMetadata {
                name: "NEAR blockchain native token",
                symbol: "NEAR",
                icon: Some(
                    "https://raw.githubusercontent.com/near/near-wallet/7ef3c824404282b76b36da2dff4f3e593e7f928d/packages/frontend/src/images/near.svg",
                ),
                decimals: 24,
            },
            block_timestamp_nanos: U64(
                1618591001077258170,
            ),
        },
        HistoryItem {
            involved_account_id: Some(
                AccountId(
                    "zubkowi.near",
                ),
            ),
            delta_balance: I128(
                -4114870673270849579420,
            ),
            balance: U128(
                31545740387248236650420580,
            ),
            cause: "TRANSACTION",
            status: "SUCCESS",
            coin_metadata: CoinMetadata {
                name: "NEAR blockchain native token",
                symbol: "NEAR",
                icon: Some(
                    "https://raw.githubusercontent.com/near/near-wallet/7ef3c824404282b76b36da2dff4f3e593e7f928d/packages/frontend/src/images/near.svg",
                ),
                decimals: 24,
            },
            block_timestamp_nanos: U64(
                1618591001077258170,
            ),
        },
        HistoryItem {
            involved_account_id: None,
            delta_balance: I128(
                3362968354850349579420,
            ),
            balance: U128(
                31549855257921507500000000,
            ),
            cause: "RECEIPT",
            status: "SUCCESS",
            coin_metadata: CoinMetadata {
                name: "NEAR blockchain native token",
                symbol: "NEAR",
                icon: Some(
                    "https://raw.githubusercontent.com/near/near-wallet/7ef3c824404282b76b36da2dff4f3e593e7f928d/packages/frontend/src/images/near.svg",
                ),
                decimals: 24,
            },
            block_timestamp_nanos: U64(
                1618590932039353161,
            ),
        },
        HistoryItem {
            involved_account_id: Some(
                AccountId(
                    "zubkowi.near",
                ),
            ),
            delta_balance: I128(
                -4114870673270849579420,
            ),
            balance: U128(
                31546412398289968250420580,
            ),
            cause: "TRANSACTION",
            status: "SUCCESS",
            coin_metadata: CoinMetadata {
                name: "NEAR blockchain native token",
                symbol: "NEAR",
                icon: Some(
                    "https://raw.githubusercontent.com/near/near-wallet/7ef3c824404282b76b36da2dff4f3e593e7f928d/packages/frontend/src/images/near.svg",
                ),
                decimals: 24,
            },
            block_timestamp_nanos: U64(
                1618590930828212179,
            ),
        },
        HistoryItem {
            involved_account_id: Some(
                AccountId(
                    "zubkowi.near",
                ),
            ),
            delta_balance: I128(
                79891276688900000000,
            ),
            balance: U128(
                31546492289566657150420580,
            ),
            cause: "CONTRACT_REWARD",
            status: "FAILURE",
            coin_metadata: CoinMetadata {
                name: "NEAR blockchain native token",
                symbol: "NEAR",
                icon: Some(
                    "https://raw.githubusercontent.com/near/near-wallet/7ef3c824404282b76b36da2dff4f3e593e7f928d/packages/frontend/src/images/near.svg",
                ),
                decimals: 24,
            },
            block_timestamp_nanos: U64(
                1618590930828212179,
            ),
        },
        HistoryItem {
            involved_account_id: None,
            delta_balance: I128(
                3362968354850349579420,
            ),
            balance: U128(
                31550527268963239100000000,
            ),
            cause: "RECEIPT",
            status: "SUCCESS",
            coin_metadata: CoinMetadata {
                name: "NEAR blockchain native token",
                symbol: "NEAR",
                icon: Some(
                    "https://raw.githubusercontent.com/near/near-wallet/7ef3c824404282b76b36da2dff4f3e593e7f928d/packages/frontend/src/images/near.svg",
                ),
                decimals: 24,
            },
            block_timestamp_nanos: U64(
                1618590906331074504,
            ),
        },
        HistoryItem {
            involved_account_id: Some(
                AccountId(
                    "zubkowi.near",
                ),
            ),
            delta_balance: I128(
                79891276688900000000,
            ),
            balance: U128(
                31547164300608388750420580,
            ),
            cause: "CONTRACT_REWARD",
            status: "FAILURE",
            coin_metadata: CoinMetadata {
                name: "NEAR blockchain native token",
                symbol: "NEAR",
                icon: Some(
                    "https://raw.githubusercontent.com/near/near-wallet/7ef3c824404282b76b36da2dff4f3e593e7f928d/packages/frontend/src/images/near.svg",
                ),
                decimals: 24,
            },
            block_timestamp_nanos: U64(
                1618590905146744810,
            ),
        },
    ],
)
